numba/_dispatcher.cpp

```diff
@@ -746,6 +759,290 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
+            // whereas
+            // PY_RETURN has `func(code: CodeType, instruction_offset: int, retval: object)`
+            // and
+            // CALL, C_RAISE, C_RETURN has `func(code: CodeType, instruction_offset: int, callable: object, arg0 object|MISSING)`
+            // i.e. the signature changes based on context. This influences the
+            // value of `nargsf` and what is wired into `callback_args`. First two
+            // arguments are always code and offset, optional third arg is
+            // the return value.
+            callback_args[0] = (PyObject*)code;
+            callback_args[1] = (PyObject*)offset_obj;
+            callback_args[2] = (PyObject*)retval;
+            PyObject ** callargs = &callback_args[0];
+
+            // finally, stage the call the the instrument
+            result = PyObject_Vectorcall(instrument, callargs, nargsf, NULL);
+
+            // decrement the tracing tracking field and set the event back to
+            // the original event (essentially, inlined
+            // PyThreadState_LeaveTracing).
+            tstate->tracing--;
+            tstate->what_event = old_what;
+
+            if (result == NULL){
+                // Error occurred in call to instrumentation.
+                Py_XDECREF(offset_obj);
+                Py_XDECREF(code);
+                return -1;
+            }
+        }
+        Py_XDECREF(offset_obj);
+        Py_XDECREF(code);
+    }
+    return 0;
+}
+
+/* invoke monitoring for PY_START if it is set */
+int static inline invoke_monitoring_PY_START(PyThreadState * tstate, Dispatcher *self) {
+    return invoke_monitoring(tstate, PY_MONITORING_EVENT_PY_START, self, NULL);
+}
+
+/* invoke monitoring for PY_RETURN if it is set */
+int static inline invoke_monitoring_PY_RETURN(PyThreadState * tstate, Dispatcher *self, PyObject * retval) {
+    return invoke_monitoring(tstate, PY_MONITORING_EVENT_PY_RETURN, self, retval);
+}
+
+/* A custom, fast, inlinable version of PyCFunction_Call() */
+static PyObject *
+call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyObject *locals)
+{
+    PyCFunctionWithKeywords fn = NULL;
+    PyThreadState *tstate = NULL;
+    PyObject * pyresult = NULL;
+
+    assert(PyCFunction_Check(cfunc));
+    assert(PyCFunction_GET_FLAGS(cfunc) == (METH_VARARGS | METH_KEYWORDS));
+    fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
+    tstate = PyThreadState_GET();
+    // issue PY_START if event is set
+    if(invoke_monitoring_PY_START(tstate, self) != 0){
+        return NULL;
+    }
+    // make call
+    pyresult = fn(PyCFunction_GET_SELF(cfunc), args, kws);
+    if (pyresult == NULL){
+        // TODO: could call monitoring RAISE callbacks here? Would have to know
+        // if the exception is a StopIteration though as that goes via a
+        // different path.
+        return NULL; // exception in Numba call, start to unwind
```

Comment:
I think we have to implement `RAISE`. Consider:

```python
# filename: profile_raise.py
import cProfile as profiler
import os
import pstats

import numpy as np

from numba import jit, types

class MyError(Exception):
    pass


def dot(a, b):
    sum = 0
    for i in range(len(a)):
        sum += a[i]*b[i]
    raise MyError
    return sum


pyfunc = dot


a = np.arange(16, dtype=np.float32)
b = np.arange(16, dtype=np.float32)
cfunc = jit((types.float32[::1], types.float32[::1],))(pyfunc)
# cfunc = pyfunc
n_calls = 123

p = profiler.Profile()
p.enable()
try:
    for _ in range(n_calls):
        try:
            cfunc(a, b)
        except MyError:
            pass
finally:
    p.disable()
stats = pstats.Stats(p).strip_dirs()
code = pyfunc.__code__
expected_key = (os.path.basename(code.co_filename),
                code.co_firstlineno,
                code.co_name,
                )
# check the cfunc (dot) call is in the stats.
assert expected_key in stats.stats
# check that the cfunc (dot) call has been made `n_calls` times.
cfunc_stats = stats.stats[expected_key]
print(cfunc_stats)

# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
# The error occurs here
# JIT prints: (1, 123, 0.000438581, 0.0005870840000000001, {('profile_raise.py', 13, 'dot'): (122, 1, 0.00034208000000000003, 0.00037650000000000004)})
# PY  prints: (123, 123, 0.001159585, 0.0011721190000000001, {})
assert cfunc_stats[:2] == (n_calls, n_calls)
```

Reply:
This should be fixed in 4f870eb. It turns out that `cProfile` is listening for a `PY_UNWIND` event associated with a function returning due to an exception being raised.